[14.0][ADD] account_move_line_accounting_description

ajaniszewska-dev opened the pull request.

ajaniszewska-dev committed:
[14.0][ADD] account_move_line_accounting_description

ajaniszewska-dev committed:
[14.0][ADD] account_move_line_accounting_description_sale

ajaniszewska-dev committed:
[14.0][ADD] account_move_line_accounting_description_purchase

ajaniszewska-dev commented on account_move_line_accounting_description_purchase/tests/test_purchase_order_description.py:
```diff
@@ -0,0 +1,76 @@
+
+        # For 1st PO check order line is same as product description
+
+        self.assertEqual(self.po_1_line.name, self.product_1.accounting_description)
```
@grindtildeath  assertion is failing here, what did i missed? tried to put pdb in` _prepare_account_move_line()` and it seems to not be called at all.

```
 File "/home/travis/build/OCA/account-invoicing/setup/account_move_line_accounting_description_purchase/odoo/addons/account_move_line_accounting_description_purchase/tests/test_purchase_order_description.py", line 68, in test_purchase_order_line_name

    self.assertEqual(self.po_1_line.name, self.product_1.accounting_description)

AssertionError: 'Virtual Interior Design' != 'Product1_acc_desc'

- Virtual Interior Design

+ Product1_acc_desc
```

grindtildeath replied:
@ajaniszewska-dev `_prepare_account_move_line` is called when you generate the vendor bill from the PO. And only on the `account.move.line` we want the accounting description to be used for name, not on `purchase.order.line` :wink:

grindtildeath commented on account_move_line_accounting_description/models/account_move.py:
```diff
@@ -0,0 +1,15 @@
+from odoo import api, fields, models
```
Missing header (on many other files too)

grindtildeath commented on account_move_line_accounting_description_sale/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "Acccount Move Accounting Sale Order Description",
+    "version": "14.0.1.0.0",
+    "summary": "Adds an 'accounting description' on sale order",
```
```suggestion
    "summary": "Consider accounting description when invoicing sale order",
```

grindtildeath commented on account_move_line_accounting_description_sale/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    "license": "AGPL-3",
+    "category": "Accounting & Finance",
+    "depends": ["sale", "product", "account_move_line_accounting_description"],
+    "data": [],
```
You could remove empty key and set module as auto install. (also for the glue to purchase module)

grindtildeath commented on account_move_line_accounting_description_sale/models/sale_order.py:
```diff
@@ -0,0 +1,13 @@
+    _inherit = "sale.order.line"
+
+    def _prepare_invoice_line(self, **optional_values):
+        line = super(SaleOrderLine, self)._prepare_invoice_line(**optional_values)
```
```suggestion
        line = super()._prepare_invoice_line(**optional_values)
```
You don't need this in py3 ;)

grindtildeath commented on account_move_line_accounting_description_sale/models/sale_order.py:
```diff
@@ -0,0 +1,13 @@
+    def _prepare_invoice_line(self, **optional_values):
+        line = super(SaleOrderLine, self)._prepare_invoice_line(**optional_values)
+        line["external_name"] = line["name"]
+        product = self.env["product.product"].search([("id", "=", line["product_id"])])
```
```suggestion
        product_id = line.get("product_id")
        if product_id:
            product = self.env["product.product"].browse(line["product_id"])
```
`browse` is better than `search` if you already have the id.
Also if the line is a note or a section `line["product_id"]` might throw a `KeyError` here.

grindtildeath commented on account_move_line_accounting_description_purchase/models/purchase_order.py:
```diff
@@ -0,0 +1,13 @@
+
+    def _prepare_account_move_line(self, move=False):
+        res = super()._prepare_account_move_line(move)
+        res["external_name"] = res["name"]
+        product = self.env["product.template"].search([("id", "=", res["product_id"])])
+        if product.accounting_description:
+            res["name"] = product.accounting_description
```
Same as above or below ;)

grindtildeath commented on account_move_line_accounting_description_purchase/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "Acccount Move Accounting Purchase Description",
+    "version": "14.0.1.0.0",
+    "summary": "Adds an 'accounting description' on purchase order",
```
To fix (cf suggestion for other module)

grindtildeath commented on account_move_line_accounting_description/views/account_move.xml:
```diff
@@ -0,0 +1,16 @@
+        <field name="model">account.move</field>
+        <field name="inherit_id" ref="account.view_move_form" />
+        <field name="arch" type="xml">
+             <xpath expr="//field[@name='invoice_line_ids']//tree" position="inside">
+                <field name="external_name" />
+            </xpath>
+            <xpath expr="//field[@name='invoice_line_ids']//form" position="inside">
+                <field name="external_name" />
+            </xpath>
```
Could you also set the name column invisible for invoice line view?

ajaniszewska-dev replied:
@grindtildeath thanks, PR updated with your remarks.

grindtildeath reviewed (changes requested):
- The change in invoice report (to use `external_name` instead of `name` is missing)
- Could you add a test for the onchange in account_move_line_accounting_description? Tip: Use Form class in tests instead of `create` to simulate the UI (it also plays onchange functions automatically)

ajaniszewska-dev commented on account_move_line_accounting_description/tests/test_account_move_line_description.py:
```diff
@@ -0,0 +1,73 @@
+
+        # self.assertEqual(
+        #     inv_line_with_product.name, inv_line_with_product.external_name
+        # )
```
@grindtildeath  any idea why `external_name` is `False` in test? (In UI it is properly set to product's name when changing product so I'd expect same here). Due to that, test is failing - same with l.71-73.

grindtildeath replied:
I would say, check with a pdb in your onchange. Ff the `name` is set to `accounting_description` if means your onchange must be played and there's probably a `False` value in the `name` beforehand on this line: `line.external_name = line.name`?

ajaniszewska-dev replied:
@grindtildeath  it wasnt saved after saving form, that's why. Changed field to related to product_id.name instead - seems to be working fine. No idea why form didnt save this field normally.

grindtildeath commented on account_move_line_accounting_description/tests/test_account_move_line_description.py:
```diff
@@ -0,0 +1,73 @@
+        super().setUpClass()
+        cls.partner_1 = cls.env.ref("base.res_partner_1")
+        cls.product_1 = cls.env.ref("product.product_product_1")
+        cls.product_2 = cls.env.ref("product.product_product_2")
```
The test would be much easier to read if you call your variables for what you expect (e.g. `cls.product_with_acc_desc` and `cls.product_without_acc_desc`) :wink:

ajaniszewska-dev committed:
fixup! [14.0][ADD] account_move_line_accounting_description

ajaniszewska-dev committed:
fixup! [14.0][ADD] account_move_line_accounting_description_sale

grindtildeath committed:
Do not set external_name as related to allow customization

grindtildeath committed:
Refactor test to be able to track issue

grindtildeath committed:
Add external_name to account.move.line_ids view

We need to display the field as invisible here although it
is already defined on invoice_line_ids tree above. Having it
defined only there would only set the external_name key with its
value inside the invoice_line_ids key of the values dictionary
that is passed to create function. However, this invoice_line_ids
key is popped in account.move._move_autocomplete_invoice_lines_create
function to keep only the line_ids to create the account.move and
avoid duplicated account.move.line records.

grindtildeath commented:
@ajaniszewska-dev Please check https://github.com/ajaniszewska-dev/account-invoicing/pull/1

ajaniszewska-dev committed:
Merge pull request #1 from grindtildeath/14.0-fix-account_move_line_accounting_description

14.0 fix account move line accounting description

kittiu commented:
@ajaniszewska-dev
Can you provide a decent readme for each module? 
Thanks!

ajaniszewska-dev committed:
Update readme

kittiu commented:
/ocabot merge nobump
